docs/manual/kernel/Vmm.tex

```diff
@@ -0,0 +1,148 @@
+
+    \draw [->, red] (vmm1) -- (hat0);
+    \path [->, in=90, out=210, red] (vmm2fudge.south) edge (hat1);
+    \draw [->, red] (vmm3) -- (hat2);
+\end{tikzpicture}
+\caption{Relationship between multiple VMMs, VM drivers and the HAT.}
+\end{figure}
+
+\subsection{Concepts}
+
+\paragraph{VM Range}
+Used internally by the virtual memory subsystem to represent a range of virtual addresses. It contains the usual base and length, as well some flags to store what the memory should be able to do. It also contains an offset field as the virtual address returned may not always point to the start of the VM range. This is to allow things like the VMM adding guard pages to the range, which are included as part of the VM range but not included in the virtual memory made available to the consumer. This also allows for virtual memory to map things in ways that might be misaligned according to the MMU, as the offset provides a byte sized granularity. Each range also contains a \verb|token| field, which holds an opaque pointer for the VM driver that is attached to this range.
+
+\paragraph{VM Driver} 
+\label{vmdrivers}
+While these are called VM \textit{drivers} they have nothing to do with the device driver subsystem, and are loaded much earlier than that. It might be more accurate to think of them as plugins to the VMM, with each one providing one type virtual memory that can be used for VM ranges.
+
+\begin{itemize}
+    \item Anon VM Driver: Provides general purpose 'working memory'. This is the most common type of virtual memory and ultimately just allocates physical memory and maps it where appropriate. This driver makes heavy use of the page fault handler to perform operations like demand paging and a primitive form copy-on-write.
+    \item VFS VM Driver: Acts as a bridge between the VMM and the VFS, and allows for mapping the contents of a file into an address space. It has full support for demand paging, and can trigger parts of a file to be loaded from disk if they are not present in the file cache.
+    \item Kernel VM Driver: This driver is a catch-all for all the other operations the kernel might need to perform on virtual memory. This driver is responsible for making sure the kernel binary is mapped properly in virtual memory, and at runtime it can be used to map MMIO (physical addresses that aren't usable memory) into an address space.
+\end{itemize}
+
+\paragraph{Virtual Memory Object}
+A \textit{virtual memory object} (VMO) is a wrapper around a VM range and the VMM that owns it. While the VMM provides all the options you could possibly need to manage virtual memory, a VMO provides only the most common ones and acts as a RAII type. The full VMM API is available can be used on memory managed by a VMO if the extra functionality is needed, but often it's not. As such a VMO is the recommended way to manage virtual memory, and you can freely pass this around the kernel as a representation of 'some memory in some address space'.
```

Comment:
i think something is odd with the punctuation here: 

"The full VMM Api is available can be used"

And what is a RAII?

Reply:
RAII is a C++ thing, basically just a way of managing object lifetimes. For a VMO it allocates virtual memory when created and frees the memory when the VMO is destroyed. So if you temporarily want some virtual memory you can put the VMO on the stack, and when that block of code ends the compiler will generate the calls to destroy the VMO, which in terms frees the virtual memory - no resource leaks!